src/zarr/storage/_cache.py

```diff
@@ -0,0 +1,532 @@
+
+        >>> import tempfile
+        >>> import zarr
+        >>> from zarr.storage import LocalStore
+        >>>
+        >>> # Create a temporary directory for the example
+        >>> temp_dir = tempfile.mkdtemp()
+        >>> store = LocalStore(temp_dir)
+        >>>
+        >>> # Create some test data first
+        >>> arr = zarr.create((1000, 1000), chunks=(100, 100), store=store, dtype='f4')
+        >>> arr[:] = 42.0
+        >>>
+        >>> # Now wrap with cache for faster access
+        >>> cached_store = zarr.LRUStoreCache(store, max_size=1024 * 1024 * 256)  # 256MB cache
+        >>> cached_arr = zarr.open(cached_store)
+        >>>
+        >>> # First access loads from disk and caches
+        >>> data1 = cached_arr[0:100, 0:100]  # Cache miss
+        >>>
+        >>> # Second access uses cache (much faster for remote stores)
+        >>> data2 = cached_arr[0:100, 0:100]  # Cache hit
+
+    For remote stores where the performance benefit is more apparent::
+
+        >>> from zarr.storage import RemoteStore
+        >>> # remote_store = RemoteStore.from_url("https://example.com/data.zarr")
+        >>> # cached_remote = zarr.LRUStoreCache(remote_store, max_size=2**28)
+
+
+    """
+
+    @property
+    def supports_writes(self) -> bool:
+        """Whether the underlying store supports write operations."""
+        return self._store.supports_writes
+
+    @property
+    def supports_deletes(self) -> bool:
+        """Whether the underlying store supports delete operations."""
+        return self._store.supports_deletes
+
+    @property
+    def supports_partial_writes(self) -> bool:
+        """Whether the underlying store supports partial write operations."""
+        return self._store.supports_partial_writes
+
+    @property
+    def supports_listing(self) -> bool:
+        """Whether the underlying store supports listing operations."""
+        return self._store.supports_listing
+
+    def __init__(self, store: Store, *, max_size: int) -> None:
+        if max_size <= 0:
+            raise ValueError("max_size must be a positive integer (bytes)")
+
+        # Always inherit read_only state from the underlying store
+        read_only = store.read_only
+        super().__init__(read_only=read_only)
+
+        self._store = store
+        self._max_size = max_size
+        self._current_size = 0
+        self._keys_cache: list[str] | None = None
+        self._contains_cache: dict[Any, Any] = {}
+        self._listdir_cache: dict[str | None, list[str]] = {}
+        self._values_cache: OrderedDict[str, bytes] = OrderedDict()
+        self._mutex = Lock()
+        self.hits = self.misses = 0
+
+        # Handle root attribute if present in underlying store
+        if hasattr(store, "root"):
+            self.root = store.root
+        else:
+            self.root = Path("/")  # Default root path
+
+    @classmethod
+    async def open(cls, store: Store, *, max_size: int, read_only: bool = False) -> "LRUStoreCache":
```

Comment:
```suggestion
    async def open(cls, store: Store, *, max_size: int, read_only: bool = False) -> Self:
```
Since this returns a instance of `cls`, the correct typing annotation here is to return `Self` (which is the type of `cls`). This means any sub-classes have the correct typing. (Although I'm not sure why one would subclass...). You'll need to import `Self` from `typing` at the top too.